lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/DelegationSpec.hs

```diff
@@ -0,0 +1,556 @@
+      --
+      -- Having a "toggle"-command instead of two separate commands, makes
+      -- generating valid arbitrary values easier.
+    | CmdAdversarialReg RewardAccount
+      -- ^ Someone could pay 2 ada to (re-)register your stake key. Your wallet
+      -- shouldn't be affected negatively from it.
+    | CmdMimicPointerOutput RewardAccount
+      -- ^ Someone could send funds to the same UTxO
+    deriving (Generic, Eq)
+
+isAdversarial :: Cmd -> Bool
+isAdversarial (CmdSetPortfolioOf _) = False
+isAdversarial (CmdAdversarialReg _) = True
+isAdversarial CmdOldWalletToggleFirstKey = False
+isAdversarial (CmdMimicPointerOutput _) = True -- Kind of
+
+instance Show Cmd where
+    show (CmdSetPortfolioOf n) = "CmdSetPortfolioOf " <> show n
+    show (CmdAdversarialReg (RewardAccount a)) = "CmdAdversarialReg " <> B8.unpack a
+    show CmdOldWalletToggleFirstKey = "CmdOldWalletToggleFirstKey"
+    show (CmdMimicPointerOutput (RewardAccount a)) = "CmdMimicPointerOutput " <> B8.unpack a
+
+instance Buildable Tx where
+    build (Tx cs [] []) = "Tx " <> listF cs
+    build (Tx cs ins outs) = "Tx " <> listF cs <> " " <> listF' (inF . fst) ins <> " -> " <> listF' outF outs
+
+inF :: TxIn -> Builder
+inF (TxIn h ix) = build h <> "." <> build (fromEnum ix)
+
+outF :: TxOut -> Builder
+outF (TxOut (Address addr) _tb) = build $ B8.unpack addr
+
+instance Buildable (DelegationState StakeKey') where
+    build ds = case state ds of
+        Zero -> "Zero"
+        One -> "One"
+        More ix p is0Reg -> blockMapF
+            [ ("nextKeyIx" :: String, build $ fromEnum ix)
+            , ("pointer", build p)
+            , ("isKey0Reg", build is0Reg)
+            ]
+
+instance Buildable PointerUTxO where
+    build (PointerUTxO i _c) = "PointerUTxO " <> inF i
+
+instance Buildable Env where
+    build env = "Env:\n" <> blockMapF
+        [ ("wallet" :: String, build $ wallet env)
+        , ("txs", blockListF' "-" txF $ reverse $ txs env)
+        , ("rejected txs", blockListF' "-" build $ reverse $ rejectedTxs env)
+        , ("ledger", build $ ledger env)
+        ]
+      where
+        txF tx = build (txid tx) <> ": " <> build tx
+
+
+rewardAccountF :: RewardAccount -> Builder
+rewardAccountF (RewardAccount a) = build (B8.unpack a)
+
+
+instance Buildable Cert where
+    build (RegisterKey k) = "RegKey " <> rewardAccountF k
+    build (Delegate k) = "Deleg " <> rewardAccountF k
+    build (DeRegisterKey k) = "DeReg " <> rewardAccountF k
+
+instance Arbitrary Cmd where
+    -- We don't want to generate too many adversarial registrations (we don't
+    -- expect them to happen in reality), but at least some, and enough to cause
+    -- consistent failures if something is wrong.
+    arbitrary = frequency
+        [ (85, CmdSetPortfolioOf . getNonNegative <$> arbitrary)
+        , (5, CmdAdversarialReg <$> arbitrary)
+        , (10, pure CmdOldWalletToggleFirstKey)
+        , (10, CmdMimicPointerOutput <$> arbitrary)
```

Comment:
85+5+10+10 = 110. Not sure if you wanted to split that across 100 :sweat_smile: ...